packages/salesforcedx-vscode-core/src/commands/baseDeployRetrieve.ts

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import {
+  getRootWorkspacePath,
+  LibraryCommandletExecutor
+} from '@salesforce/salesforcedx-utils-vscode/out/src';
+import {
+  Row,
+  Table
+} from '@salesforce/salesforcedx-utils-vscode/out/src/output';
+import { ContinueResponse } from '@salesforce/salesforcedx-utils-vscode/out/src/types';
+import {
+  ComponentSet,
+  DeployResult,
+  MetadataComponent,
+  MetadataResolver,
+  registryData,
+  RetrieveResult as MetadataApiRetrieveResult,
+  SourceRetrieveResult,
+  ToolingApi
+} from '@salesforce/source-deploy-retrieve';
+import {
+  ComponentStatus,
+  RequestStatus
+} from '@salesforce/source-deploy-retrieve/lib/src/client/types';
+import { join, sep } from 'path';
+import { BaseDeployExecutor } from '.';
+import { channelService, OUTPUT_CHANNEL } from '../channels';
+import { workspaceContext } from '../context';
+import { handleDeployDiagnostics } from '../diagnostics';
+import { nls } from '../messages';
+import { DeployQueue } from '../settings';
+import { SfdxPackageDirectories, SfdxProjectConfig } from '../sfdxProject';
+
+type RetrieveResult = MetadataApiRetrieveResult | SourceRetrieveResult;
+type DeployRetrieveResult = DeployResult | RetrieveResult;
+
+export abstract class DeployRetrieveExecutor<
+  T
+> extends LibraryCommandletExecutor<T> {
+  constructor(executionName: string, logName: string) {
+    super(executionName, logName, OUTPUT_CHANNEL);
+  }
+
+  public async run(response: ContinueResponse<T>): Promise<boolean> {
+    let result: DeployRetrieveResult | undefined;
+
+    try {
+      const components = await this.getComponents(response);
+
+      this.telemetry.addProperty(
+        'metadataCount',
+        this.createComponentCount(components)
+      );
+
+      result = await this.doOperation(components);
+
+      const status = this.getStatus(result);
+
+      return (
+        status === RequestStatus.Succeeded ||
+        status === RequestStatus.SucceededPartial
+      );
```

Comment:
All DeployRetrieve commands need to report metadata counts and report status to the base executor.